Refactor task configs.

kevinzakka opened the pull request:
**TLDR:** This is a structural refactor of how task configs are built. The actual task logic (rewards, observations, actions, etc.) is unchanged.

## Context

We initially used dataclasses with `__post_init__`, but this had fundamental issues:
- `__post_init__` silently overrides constructor arguments
- CLI arguments like `--env.episode-length-s 10.0` get ignored
- Field redeclaration was verbose and error-prone

We refactored to a function-based approach, but it had its own problems:
- Base factory took 10+ robot-specific parameters
- Robot configs had to pass all params upfront (even for defaults)
- Configs were singletons (via `@retval` decorator)
- Creating variants required `deepcopy()` everywhere

## This PR

Simple factory pattern: **base factory creates structure, robot configs mutate and return**.

Base factory defines the task structure:
```python
def make_velocity_env_cfg() -> ManagerBasedRlEnvCfg:
  """Creates base config with placeholder values."""
  # Build observations, actions, rewards, etc.
  return ManagerBasedRlEnvCfg(...)
```

Robot configs customize as needed:
```python
def unitree_go1_rough_env_cfg(play: bool = False) -> ManagerBasedRlEnvCfg:
  cfg = make_velocity_env_cfg()

  # Customize robot-specific parts
  cfg.scene.entities["robot"] = get_go1_robot_cfg()
  cfg.actions["joint_pos"].scale = GO1_ACTION_SCALE
  cfg.rewards["pose"].params["std_standing"] = {...}

  if play:
    cfg.episode_length_s = int(1e9)
    cfg.observations["policy"].enable_corruption = False

  return cfg
```

## What changed

- Base factory takes zero parameters (no more parameter explosion)
- Eliminated `@retval` hack and `deepcopy()` boilerplate
- No more `__post_init__` footguns - mutations are explicit
- Customization is simple: call base factory, mutate what you need, return
- Registry now accepts functions instead of instances (lazy evaluation)
- Play mode is now a `play: bool` parameter instead of duplicate config definitions

## Testing

This refactor also introduces 3 new test files that check that:
- All configs load without errors
- Play/training modes are correctly configured
- Observation corruption enabled/disabled appropriately
- Config structures match between play and training variants

The mutation-based approach is now validated by comprehensive tests, so silent failures get caught.

## Tradeoffs

**What we lose:**
- Mutation-based customization (`cfg.rewards["pose"].params["std_standing"] = {...}`) doesn't give autocomplete or type checking for nested dict keys
- More imperative than declarative (compared to dataclass field definitions)

**What we gain:**
- Mutations are explicit and visible (no `__post_init__` magic)
- Don't need to predict all customization points upfront
- Simpler to understand - no complex inheritance hierarchy to trace
- Tests validate the mutation paths, catching silent failures

Overall, this is a better tradeoff: we give up some type safety at the mutation sites, but gain clarity, flexibility, and comprehensive test coverage. The explicit nature of mutations makes the code easier to understand and modify.

brentyi commented on src/mjlab/tasks/velocity/velocity_env_cfg.py:
```diff
@@ -252,7 +195,7 @@ def create_velocity_env_cfg(
       params={
         "std": math.sqrt(0.2),
-        "asset_cfg": SceneEntityCfg("robot", body_names=(viewer_body_name,)),
+        "asset_cfg": SceneEntityCfg("robot", body_names=()),  # Set per-robot.
```
I agree aesthetically that large signatures for functions like `create_velocity_env_cfg() ` were unideal, but this "Set per-robot" as a comment does feel more brittle to me... like if the "base" config changes to require more fields that need to be overwritten, it's much easier to have silent failures

brentyi commented on src/mjlab/scripts/demo.py:
```diff
@@ -29,10 +29,9 @@ def main() -> None:
     ),
   )
-  run_play("Mjlab-Tracking-Flat-Unitree-G1", args)
+  run_play("Mjlab-Tracking-Flat-Unitree-G1-Demo", args)
```
Since the demo is a one-off thing and the only difference is sampling mode, I still kind of prefer just loading the play config and then overwriting the sampling mode

brentyi reviewed (approved):
Looks much better!!

My weak feeling is that it would be nice to drop the `Callable[[], ManagerBasedRlEnvCfg]` from the task registration signature for simplicity; config construction is very fast and forcing all of the factories to run might surface bugs/errors faster.

If every task expects a play variant, perhaps we could just include the play config as part of the task definition?
```python
  env_cfg: ManagerBasedRlEnvCfg,
  play_env_cfg: ManagerBasedRlEnvCfg,
  rl_cfg: RslRlOnPolicyRunnerCfg,
```

kevinzakka commented:
@brentyi Adopted both suggestions! Removed `Callable` and added `play_env_cfg` parameter to group training/play configs in a single registration. Also discovered and fixed a pre-existing bug where terrain generators were being shared across configs.
